File: code/reporter/src/app/db/models/target.py
from datetime import datetime


from app.db.models.base_model import BaseDocumentModel
from app.enums.main import ScanStatus
from app.utils import ip_utils
from zoneinfo import ZoneInfo


class Target(BaseDocumentModel):
    customer_username: str
    target_address: str
    target_type: str
    flag: int = 0
    scan_status: ScanStatus.ENUM
    created_at: datetime
    is_ds: bool = False
    scans: list | None = None
    scanner_ip: str | None = None
    scanner_username: str | None = None
    scan_started_time: datetime | None = None
    scan_completed_time: datetime | None = None
    overall_cvss_score: float | None = None
    cvss_score_by_host: dict | None = None
    failure_reason: int | None = None

    def is_ip_range(self):
        if ip_utils.is_valid_cidr_by_regex(self.target_address):
            return True

    def get_scan_started_time(self):
        if not self.scan_started_time:
            return ""

        # Assuming scan_started_time is stored in UTC, we ensure it's timezone-aware
        if self.scan_started_time.tzinfo is None:
            # If it's naive, assume it's in UTC
            self.scan_started_time = self.scan_started_time.replace(
                tzinfo=ZoneInfo("UTC")
            )
            print(f"Tzinfo is None: {self.scan_started_time}")
        print(f"Scan Started Time: {self.scan_started_time}")

        # Convert from UTC to the system's local timezone
        local_timezone = ZoneInfo("localtime")  # Uses the system's local timezone
        scan_started_time_local = self.scan_started_time.astimezone(local_timezone)

        return scan_started_time_local.strftime("%d-%m-%Y %I:%M%p")

    def get_scan_completed_time(self):
        if not self.scan_completed_time:
            return ""

        # Assuming scan_completed_time is stored in UTC, we ensure it's timezone-aware
        if self.scan_completed_time.tzinfo is None:
            # If it's naive, assume it's in UTC
            self.scan_completed_time = self.scan_completed_time.replace(
                tzinfo=ZoneInfo("UTC")
            )
            print(f"Tzinfo is None: {self.scan_completed_time}")
        print(f"Scan Completed Time: {self.scan_completed_time}")

        # Convert from UTC to the system's local timezone
        local_timezone = ZoneInfo("localtime")  # Uses the system's local timezone
        scan_completed_time_local = self.scan_completed_time.astimezone(local_timezone)

        return scan_completed_time_local.strftime("%d-%m-%Y %I:%M%p")

File: code/scanner/models/target.go
package models

import (
	"regexp"
	"time"

	"github.com/CSPF-Founder/iva/scanner/enums"
	"go.mongodb.org/mongo-driver/bson/primitive"
)

type ScanInfo struct {
	ScanNumber int       `bson:"scan_number"`
	ScanDate   time.Time `bson:"scan_date"`
}
type Target struct {
	ID               primitive.ObjectID `bson:"_id"`
	CustomerUsername string             `bson:"customer_username"`
	TargetAddress    string             `bson:"target_address"`

	Flag       int                `bson:"flag"`
	ScanStatus enums.TargetStatus `bson:"scan_status"`
	CreatedAt  time.Time          `bson:"createdAt"`
	IsDS       bool               `bson:"is_ds"`
	Scans      []ScanInfo         `bson:"scans"`

	TargetType        enums.TargetType `bson:"target_type"`
	ScannerIP         string           `bson:"scanner_ip"`
	ScannerUsername   string           `bson:"scanner_username"`
	ScanStartedTime   time.Time        `bson:"scan_started_time"`
	ScanCompletedTime time.Time        `bson:"sscan_completed_time"`
	FailureReason     int              `bson:"failure_reason"`

	OverallCVSSScore *float64
	CVSSScoreByHost  map[string]float64
}

func (s *Target) IsIPRange() bool {
	ipRangeRegex := regexp.MustCompile(`^\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}/\d{1,2}$`)
	return ipRangeRegex.MatchString(s.TargetAddress)
}

// GetNextScanNumber returns the next scan number for the target
func (s *Target) GetNextScanNumber() int {
	if len(s.Scans) == 0 {
		return 1
	}

	bigScanNum := 1
	for _, scan := range s.Scans {
		if scan.ScanNumber > bigScanNum {
			bigScanNum = scan.ScanNumber
		}
	}

	return bigScanNum + 1
}
